File: app/src/main/java/com/constraint/vagabond/main/DetailsActivity.java
package com.constraint.vagabond.main;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuInflater;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.constraint.vagabond.R;
import com.constraint.vagabond.data.DataStore;
import com.constraint.vagabond.data.RecreationalAreaList;
import com.squareup.picasso.Picasso;

import java.util.List;

public class DetailsActivity extends AppCompatActivity implements DetailsContract {

  private CollapsingToolbarLayout collapsingToolbarLayout;
  private Toolbar toolbar;
  private FloatingActionButton favButton;
  private String collapsingImageUrl;
  private TextView areaDescription;
  private RecreationalAreaList detailedArea = DataStore.getRecreationalAreaList();
  private List<String> imageUrls;
  private int position;
  private RecyclerView areaImages;
  private Button directionBtn;

  @Override
  protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    Intent getData = getIntent();
    collapsingImageUrl = getData.getStringExtra("image_url");
    position = getData.getIntExtra("position", 0);
    imageUrls = detailedArea.recreationalAreaList.get(position).getImageUrls();
    initializeView();
  }

  @Override
  protected void onStart() {
    super.onStart();
    initializeData();
    loadImage();
    setDataToRecyclerView();
  }

  @Override
  public void initializeView() {
    setContentView(R.layout.activity_detail);
    toolbar = findViewById(R.id.toolbar_title);
    setSupportActionBar(toolbar);
    getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    collapsingToolbarLayout = findViewById(R.id.collapsing_layout);
    areaDescription = findViewById(R.id.description);
    favButton = findViewById(R.id.fav_button);
    areaImages = findViewById(R.id.horizontal_recycler_view);
    areaDescription = findViewById(R.id.description);
    directionBtn = findViewById(R.id.directions);
  }

  public void loadImage() {
    final ImageView expandingImage = findViewById(R.id.collapsing_image);
    Picasso.get().load(collapsingImageUrl).fit().centerCrop().into(expandingImage);
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    MenuInflater inflator = getMenuInflater();
    inflator.inflate(R.menu.menu_main, menu);
    return true;
  }

  @Override
  public void destroyView() {}

  @Override
  public void initializeData() {
    collapsingToolbarLayout.setTitle(detailedArea.recreationalAreaList.get(position).recAreaName);
    String desc = detailedArea.recreationalAreaList.get(position).recAreaDescription;
    areaDescription.setText(desc);
  }

  @Override
  public void setDataToRecyclerView() {
    LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
    areaImages.setLayoutManager(linearLayoutManager);
    DetailsPhotosAdapter detailsPhotosAdapter = new DetailsPhotosAdapter(imageUrls);
    areaImages.setAdapter(detailsPhotosAdapter);
  }
}

File: app/src/main/java/com/sierra/vagabond/main/details/DetailsActivity.kt
package com.sierra.vagabond.main.details

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.sierra.vagabond.R
import com.sierra.vagabond.data.RecAreaRepository
import com.sierra.vagabond.data.entities.RecreationalArea
import com.sierra.vagabond.main.details.adapter.DetailsPhotosAdapter
import com.sierra.vagabond.main.details.adapter.FacilitiesAdapter
import com.sierra.vagabond.utils.REC_AREA_ID
import com.squareup.picasso.Picasso
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_detail.*

class DetailsActivity : AppCompatActivity(), DetailsMvp.View {

    private lateinit var detailedArea: RecreationalArea
    private val presenter: DetailsPresenter = DetailsPresenter(this)
    private lateinit var repo: RecAreaRepository
    private lateinit var id: String
    private val compositeDisposable: CompositeDisposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail)
//        detailedArea = intent.getParcelableExtra(AREA)
        id = intent.getStringExtra(REC_AREA_ID)
        repo = RecAreaRepository(this)
        getAreaFromRepo()
        setSupportActionBar(toolbar_title)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun getAreaFromRepo() {
        val areaDisposable = repo.getSingleArea(id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { gotArea ->
                            detailedArea = gotArea
                            initializeData()
                            loadImage()
                            setFacilities()
                        },
                        { err -> Log.d(javaClass.simpleName, "Null result") }
                )
        compositeDisposable.add(areaDisposable)
    }

    private fun loadImage() {
        val expandingImage = findViewById<ImageView>(R.id.collapsing_image)
        Picasso.get().load(detailedArea.recAreaMediaList[0].imageURL).fit().centerCrop().into(expandingImage)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }

    private fun initializeData() {
        collapsing_layout.title = detailedArea.recAreaName
        val desc = detailedArea.recAreaDescription
        description.text = desc
    }

    private fun setFacilities() {
        val linearLayoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        list_facilities.layoutManager = linearLayoutManager
        val facilities = detailedArea.recAreaFacilities
        val facilitiesAdapter = FacilitiesAdapter(facilities)
        list_facilities.adapter = facilitiesAdapter
    }

    override fun setDataToRecyclerView(imageUrls: List<String>) {
        val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        horizontal_recycler_view.layoutManager = linearLayoutManager
        val detailsPhotosAdapter = DetailsPhotosAdapter(imageUrls)
        horizontal_recycler_view.adapter = detailsPhotosAdapter
    }
}
